File: genql_ruby/lib/genql/compilers/sql_compiler.rb
# genql/compilers/sql_compiler
#
# Compiles DATA intents into SQL statements.
#
# Supported body keys
# -------------------
# operation : String  — "select" | "insert" | "update" | "delete"
# table     : String  — target table name
# columns   : Array   — columns to retrieve (select only); omit or ["*"] for all
# where     : Hash    — equality conditions, e.g. {id: 42}
# values    : Hash    — column → value mapping (insert / update)

require_relative "base"
require_relative "../intent"

module Genql
  module Compilers
    class SQLCompiler < Compiler
      def target_language
        "sql"
      end

      def supported_kinds
        [IntentKind::DATA]
      end

      def compile(intent)
        op = (body_get(intent.body, "operation") || "select").to_s.downcase
        case op
        when "select" then sql_select(intent.body)
        when "insert" then sql_insert(intent.body)
        when "update" then sql_update(intent.body)
        when "delete" then sql_delete(intent.body)
        else
          raise ArgumentError,
                "SQLCompiler: unknown operation #{op.inspect}. " \
                "Expected one of [\"select\", \"insert\", \"update\", \"delete\"]."
        end
      end

      private

      # Accept both string and symbol keys
      def body_get(body, key)
        body[key.to_sym] || body[key.to_s]
      end

      def quote(identifier)
        "\"#{identifier}\""
      end

      def literal(value)
        case value
        when String then "'#{value.gsub("'", "''")}'"
        when NilClass then "NULL"
        else value.to_s
        end
      end

      def where_clause(where)
        return "" if where.nil? || where.empty?

        parts = where.map { |col, val| "#{quote(col)} = #{literal(val)}" }
        " WHERE #{parts.join(' AND ')}"
      end

      def sql_select(body)
        table   = quote(body_get(body, "table"))
        cols    = body_get(body, "columns") || ["*"]
        col_str = cols.map { |c| c == "*" ? "*" : quote(c) }.join(", ")
        where   = where_clause(body_get(body, "where") || {})
        "SELECT #{col_str} FROM #{table}#{where};"
      end

      def sql_insert(body)
        table  = quote(body_get(body, "table"))
        values = body_get(body, "values") || {}
        cols   = values.keys.map { |c| quote(c) }.join(", ")
        vals   = values.values.map { |v| literal(v) }.join(", ")
        "INSERT INTO #{table} (#{cols}) VALUES (#{vals});"
      end

      def sql_update(body)
        table  = quote(body_get(body, "table"))
        values = body_get(body, "values") || {}
        set    = values.map { |c, v| "#{quote(c)} = #{literal(v)}" }.join(", ")
        where  = where_clause(body_get(body, "where") || {})
        "UPDATE #{table} SET #{set}#{where};"
      end

      def sql_delete(body)
        table = quote(body_get(body, "table"))
        where = where_clause(body_get(body, "where") || {})
        "DELETE FROM #{table}#{where};"
      end
    end
  end
end

File: genql/compilers/sql_compiler.py
"""
genql.compilers.sql_compiler
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compiles DATA intents into SQL statements.

Supported body keys
-------------------
operation : str
    ``"select"`` | ``"insert"`` | ``"update"`` | ``"delete"``
table : str
    Target table name.
columns : list[str]  (select only)
    Columns to retrieve; omit or pass ``["*"]`` for all.
where : dict[str, Any]  (select / update / delete)
    Equality conditions, e.g. ``{"id": 42}``.
values : dict[str, Any]  (insert / update)
    Column → value mapping.
"""

from __future__ import annotations

from typing import Any, Dict, List

from genql.compilers.base import Compiler
from genql.intent import Intent, IntentKind


class SQLCompiler(Compiler):
    """Translates :attr:`IntentKind.DATA` intents to SQL."""

    @property
    def target_language(self) -> str:
        return "sql"

    @property
    def supported_kinds(self) -> List[IntentKind]:
        return [IntentKind.DATA]

    # ------------------------------------------------------------------ #
    # Public API                                                           #
    # ------------------------------------------------------------------ #

    def compile(self, intent: Intent) -> str:
        op = str(intent.body.get("operation", "select")).lower()
        dispatch = {
            "select": self._select,
            "insert": self._insert,
            "update": self._update,
            "delete": self._delete,
        }
        if op not in dispatch:
            raise ValueError(
                f"SQLCompiler: unknown operation {op!r}. "
                f"Expected one of {list(dispatch)!r}."
            )
        return dispatch[op](intent.body)

    # ------------------------------------------------------------------ #
    # Private helpers                                                      #
    # ------------------------------------------------------------------ #

    def _quote(self, identifier: str) -> str:
        return f'"{identifier}"'

    def _literal(self, value: Any) -> str:
        if isinstance(value, str):
            escaped = value.replace("'", "''")
            return f"'{escaped}'"
        if value is None:
            return "NULL"
        return str(value)

    def _where_clause(self, where: Dict[str, Any]) -> str:
        if not where:
            return ""
        parts = [
            f"{self._quote(col)} = {self._literal(val)}"
            for col, val in where.items()
        ]
        return " WHERE " + " AND ".join(parts)

    def _select(self, body: Dict[str, Any]) -> str:
        table = self._quote(body["table"])
        cols = body.get("columns") or ["*"]
        col_list = ", ".join(
            "*" if c == "*" else self._quote(c) for c in cols
        )
        where = self._where_clause(body.get("where") or {})
        return f"SELECT {col_list} FROM {table}{where};"

    def _insert(self, body: Dict[str, Any]) -> str:
        table = self._quote(body["table"])
        values: Dict[str, Any] = body.get("values") or {}
        cols = ", ".join(self._quote(c) for c in values)
        vals = ", ".join(self._literal(v) for v in values.values())
        return f"INSERT INTO {table} ({cols}) VALUES ({vals});"

    def _update(self, body: Dict[str, Any]) -> str:
        table = self._quote(body["table"])
        values: Dict[str, Any] = body.get("values") or {}
        set_clause = ", ".join(
            f"{self._quote(c)} = {self._literal(v)}" for c, v in values.items()
        )
        where = self._where_clause(body.get("where") or {})
        return f"UPDATE {table} SET {set_clause}{where};"

    def _delete(self, body: Dict[str, Any]) -> str:
        table = self._quote(body["table"])
        where = self._where_clause(body.get("where") or {})
        return f"DELETE FROM {table}{where};"
